### PR TITLE
refactor(tty): call setup_console after pivot_root, use syscall for mount_console

### PR DESCRIPTION
**Pull request by nayuta723**
## Description

CRIU checkpoint fails for TTY-enabled containers because the PTY belongs to the host's devpts, not the container's.
CRIU requires that all FDs can be traced back to mounts **within the container's mount namespace**.

**Solution**

Align `setup_console` with runc's approach: 

1. Call `setup_console` after `pivot_root`
   - PTY is created from container's `/dev/pts/ptmx`, not host's
   - FDs now belong to a mount CRIU can track

2. Unify init/exec console setup
   - `setup_console(syscall, fd, mount)` with `mount` param (like runc)
   - `mount=true` for init, `false` for exec

3. Use `syscall.mount()` for consistency

**Result**

| Before | After |
|--------|-------|
| FD 0 → host's `/dev/pts/x` | FD 0 → container's `/dev/pts/0` |
| CRIU: mount not found | CRIU: checkpoint success |

**Reference**
- https://github.com/opencontainers/runc/blob/v1.4.0/libcontainer/standard_init_linux.go
- https://github.com/opencontainers/runc/blob/v1.4.0/libcontainer/rootfs_linux.go

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Test updates
- [ ] CI/CD related changes
- [ ] Other (please describe):

## Testing

- [ ] Added new unit tests
- [ ] Added new integration tests
- [x] Ran existing test suite
- [x] Tested manually (please provide steps)

**Setup**
```
sudo podman run -d --name test-youki-tty --runtime $RUNTIME -t alpine sh -c 'i=0; while true; do echo $i; i=$((i+1)); sleep 1; done' 2>&1
```

**Result**
- youki（main branch）
```
=== YOUKI (main branch) Container ===
--- FD links ---
total 0
dr-x------ 2 root root  3 Dec 21 05:03 .
dr-xr-xr-x 9 root root  0 Dec 21 05:03 ..
lrwx------ 1 root root 64 Dec 21 05:04 0 -> /dev/pts/8
lrwx------ 1 root root 64 Dec 21 05:04 1 -> /dev/pts/8
lrwx------ 1 root root 64 Dec 21 05:04 2 -> /dev/pts/8
--- mountinfo devpts ---
546 544 0:101 / /dev/pts rw,nosuid,noexec,relatime - devpts devpts rw,gid=5,mode=620,ptmxmode=666

```
- runc
```
=== RUNC Container ===
--- FD links ---
total 0
dr-x------ 2 root root  3 Dec 21 04:58 .
dr-xr-xr-x 9 root root  0 Dec 21 04:58 ..
lrwx------ 1 root root 64 Dec 21 04:58 0 -> /dev/pts/0
lrwx------ 1 root root 64 Dec 21 04:58 1 -> /dev/pts/0
lrwx------ 1 root root 64 Dec 21 04:58 2 -> /dev/pts/0
--- mountinfo devpts ---
509 507 0:94 / /dev/pts rw,nosuid,noexec,relatime - devpts devpts rw,gid=5,mode=620,ptmxmode=666
434 507 0:94 /0 /dev/console rw,nosuid,noexec,relatime - devpts devpts rw,gid=5,mode=620,ptmxmode=666
```
- youki（This PR）
```
=== YOUKI Container ===
--- FD links ---
total 0
dr-x------ 2 root root  3 Dec 21 04:51 .
dr-xr-xr-x 9 root root  0 Dec 21 04:51 ..
lrwx------ 1 root root 64 Dec 21 04:51 0 -> /dev/pts/0
lrwx------ 1 root root 64 Dec 21 04:51 1 -> /dev/pts/0
lrwx------ 1 root root 64 Dec 21 04:51 2 -> /dev/pts/0
--- mountinfo devpts ---
456 454 0:66 / /dev/pts rw,nosuid,noexec,relatime - devpts devpts rw,gid=5,mode=620,ptmxmode=666
379 454 0:66 /0 /dev/console rw,nosuid,noexec,relatime - devpts devpts rw,gid=5,mode=620,ptmxmode=666
```

## Related Issues

related to #2335 

## Additional Context

I have verified that the checkpoint functions correctly, as demonstrated by the tests in this Draft PR: https://github.com/nayuta723/youki/pull/6

### CODE REVIEW

File: crates/libcontainer/src/tty.rs [hunk 1]
```diff
@@ -120,34 +141,128 @@ pub fn setup_console_socket(
...
+///
+/// # Arguments
+/// * `console_fd` - The console socket file descriptor
+/// * `mount` - Whether to mount PTY slave on /dev/console (true for init, false for exec)
+///
+/// By creating PTY from container's devpts, the PTY belongs to a mount that
+/// exists within the container's namespace, which is required for CRIU checkpoint.
+///
+/// See: https://github.com/opencontainers/runc/blob/v1.4.0/libcontainer/rootfs_linux.go
+pub fn setup_console(syscall: &dyn Syscall, console_fd: RawFd, mount: bool) -> Result<()> {
```
Comment (nayuta723) [inline crates/libcontainer/src/tty.rs:286]: Since I'm changing the arguments of a public function, should I keep the original function as deprecated and add a new one instead to avoid a breaking change?

Comment (utam0k) [reply crates/libcontainer/src/tty.rs:286]: No need. We'll release v0.6.0 in the next version, so we're not too concerned about breaking changes.

**Comment by nayuta723** [comment]
@utam0k 
> This change addresses https://github.com/advisories/GHSA-qw9x-cqr3-wc7r. Please verify its safety.

What is the recommended approach for verifying this fix? I want to ensure we have adequate test coverage for this specific edge case.

**Comment by utam0k** [comment]
@nayuta723 As far as I know, it's hard to cover this with e2e tests because it would be caused by TOCTOU. The most common safety method is to add verification and use [mount_from_fd](https://github.com/youki-dev/youki/blob/93951d2bedd732e8a02e4976a486b080a519379a/crates/libcontainer/src/syscall/syscall.rs#L48). 
runc's code:
https://github.com/opencontainers/runc/commit/2f74f4ae1b483b3a4d197f574a37c7b768baf96f#diff-af76b9486c10ca0b6b87e8b8229fa12da73d8ece65e42a0358bec25bfbc21cd9

For example, it might be possible to test this by placing a malicious symbolic link within the rootfs from the start, but it might not be that simple.

**Comment by nayuta723** [comment]
Thanks. In that case, can I assume that the goal of this PR is to demonstrate that the CVE-2025-52565 vulnerability can no longer be reproduced? I will set up a reproduction environment and verify that the updated implementation effectively prevents the exploit.

**Comment by utam0k** [comment]
> Thanks. In that case, can I assume that the goal of this PR is to demonstrate that the [CVE-2025-52565](https://github.com/advisories/GHSA-qw9x-cqr3-wc7r) vulnerability can no longer be reproduced? I will set up a reproduction environment and verify that the updated implementation effectively prevents the exploit.

The goal is not only to reproduce it, but also to avoid directly using `mount`. In other words, we aim to verify the filepath (file descriptor) before mounting and then mount the path using the file descriptor.

**Comment by nayuta723** [comment]
@utam0k 

To mitigate TOCTOU (Time-of-Check to Time-of-Use) attacks, I have introduced `verify_ptmx_handle` and `verify_pty_slave` to validate the file descriptors of the link targets. I have also added unit tests to cover cases where the targets are not a PTMX or PTY:

Valid cases:

- `test_verify_pty_slave_with_real_pty`
- `test_verify_ptmx_handle_with_real_pty`

Invalid cases:

- `test_verify_ptmx_handle_with_regular_file`
- `test_verify_pty_slave_with_regular_file`

Furthermore, I have updated `mount_console` to use `mount_from_fd`. This ensures the operation remains secure even if the path is replaced by a symbolic link.

Do you think these tests are enough, or should I try any other specific cases to make sure the fix is complete?

File: crates/libcontainer/src/tty.rs [hunk 2]
```diff
@@ -120,34 +144,261 @@ pub fn setup_console_socket(
...
+#[cfg(not(target_env = "musl"))]
+const DEVPTS_SUPER_MAGIC: i64 = 0x1cd1;
+
+/// Verify file descriptor using stat and statfs, similar to runc's VerifyInode.
+///
+/// This is a helper function that gets stat/statfs for a file descriptor and
+/// calls the provided verification function with the results.
+///
+/// Ref: https://github.com/opencontainers/runc/blob/v1.4.0/libcontainer/system/linux.go
+fn verify_inode<F>(fd: &OwnedFd, verify: F) -> Result<()>
```
Comment (utam0k) [inline crates/libcontainer/src/tty.rs]: How about making it util function not only for tty?

File: crates/libcontainer/src/tty.rs [hunk 3]
```diff
@@ -120,34 +144,261 @@ pub fn setup_console_socket(
...
+        if let Err(err) = mount_console(syscall, slave) {
+            tracing::warn!(
+                ?err,
+                "failed to mount /dev/console, CRIU checkpoint may not work"
+            );
+        }
+    }
+
+    // Send PTY master to console socket
+    let pty_name: &[u8] = b"/dev/ptmx";
```
Comment (utam0k) [inline crates/libcontainer/src/tty.rs]: It can modify a constant.

File: crates/libcontainer/src/tty.rs [hunk 4]
```diff
@@ -120,34 +144,261 @@ pub fn setup_console_socket(
...
+    verify_ptmx_handle(master)?;
+    verify_pty_slave(slave)?;
+
+    // Mount PTY slave on /dev/console (only for init container)
+    if mount {
+        if let Err(err) = mount_console(syscall, slave) {
+            tracing::warn!(
+                ?err,
+                "failed to mount /dev/console, CRIU checkpoint may not work"
+            );
```
Comment (utam0k) [inline crates/libcontainer/src/tty.rs]: Could you explain why we should skip this error

**Comment by utam0k** [comment]
> Do you think these tests are enough, or should I try any other specific cases to make sure the fix is complete?

They are addressing our needs.

File: crates/libcontainer/src/tty.rs (hunk 4 above)
Comment (nayuta723) [reply crates/libcontainer/src/tty.rs]: Fixed the missing parts. Sorry for the oversight.
